stan/math/prim/functor/reduce_sum.hpp

```diff
@@ -20,76 +20,113 @@ namespace math {
 
 namespace internal {
 
+/**
+ * Base definition of implimentation called by reduce `reduce_sum`.
+ * @tparam ReduceFunction An type with a valid `operator()`
+ * @tparam ReturnType The return type of the call to
+ * `ReduceFunction::operator()`
+ * @tparam Vec type of the first input argument with an `operator[]`
+ * @tparam Args Parameter pack holding the types of the object send to
+ * `ReduceFunction::operator()`
+ */
 template <typename ReduceFunction, typename Enable, typename ReturnType,
           typename Vec, typename... Args>
 struct reduce_sum_impl {
-  return_type_t<Vec, Args...> operator()(const Vec& vmapped,
-                                         std::size_t grainsize,
-                                         std::ostream* msgs,
-                                         const Args&... args) const {
+  template <typename OpVec, typename... OpArgs,
```

Comment:
Why template on OpVec and OpArgs if we enforce that they be the same types as Vec and Args?

Reply:
So we could get rid of them, having them allows the value type of `Vec` and `OpVec` to differ. (i.e. `Vec` could be `std::vector<T>&` and `OpVec` could be `std::vector<T>&&`). Removing them means that we couldn't have the value types in the template of the class differ from the constructor's templates. But idt that would be a big deal

Reply:
Given we do the calls like:

```
internal::reduce_sum_impl<ReduceFunction, void, return_type, Vec,
                                   Args...>()(
      std::forward<Vec>(vmapped), grainsize, msgs, std::forward<Args>(args)...)
```

I don't think they're ever different. It makes me wonder if reduce_sum_impl should just be a function.

Reply:
It has no member variables, and we're only calling operator() right after construction.

Reply:
I'm afk but theres a link in the docs to the tbb site and I think reduce sum needs to be a struct/class

Reply:
The reason I went this way was to have the template magic pick the right version. This template magic does not work so well for functions as I recall.

@SteveBronder it's right that parallel_reduce needs a struct, but you could define that within the body of a function if you want to.

Somehow I prefer the struct; feels safer... but whatever works.

Reply:
In the previous version there were member variables if I recall correctly. I think there was a thread_local pointer to copies of what were the shared arguments or something?

Reply:
Yes, I tried that, but for the moment I do not see a safe way to make sure things stay (and get cleaned) from the thread local storage thing. I tried on the v6 branch to introduce a new `local_vari` thing for the deep copies. This is probably the way to go at some later point; right now I am still getting failed unit tests on that branch for reasons not so clear to me now... this can be revisited later.